R/zchunk_LA132.industry.R

```diff
@@ -48,48 +48,160 @@ module_energy_LA132.industry_DISABLED <- function(command, ...) {
+      filter(fuel %in% c("gas", "coal")) ->
+      L132.in_EJ_R_indrefining_F_Yh
+
+    # CHP: no adjustments necessary
+    L123.in_EJ_R_indchp_F_Yh -> L132.in_EJ_R_indchp_F_Yh
+
+    # Combine all of the deduction tables and multiply by -1 to indicate that these are deductions
+    L132.in_EJ_R_indunoil_F_Yh %>%
+      bind_rows(L132.in_EJ_R_indrefining_F_Yh) %>%
+      bind_rows(L132.in_EJ_R_indgasproc_F_Yh) %>%
+      bind_rows(L132.in_EJ_R_indchp_F_Yh) %>%
+      mutate(value = -1 * value) ->
+      L132.in_EJ_R_inddeductions_F_Yh
+
+    ## Heat: fuel inputs to heat need to be added to industrial energy use, in regions where heat is not modeled as a final fuel
+    # Calculate the share of heat consumed by the industrial sector, in regions where heat is not modeled as a separate fuel
+    L131.share_R_Senduse_heat_Yh %>%
+      filter(grepl("industry", sector)) %>%
+      select(-sector) %>%
+      group_by(GCAM_region_ID, fuel, year) %>%
+      summarise(value = sum(value)) %>%
+      ungroup ->
+      L132.share_R_indenergy_heat_Yh
+
+    # Multiply these shares by the energy inputs to heat
+    A_regions %>%
+      filter(has_district_heat == 0) %>%
+      select(GCAM_region_ID) %>%
+      unlist ->
+      has_district_heat_GCAM_region_ID
+
+    L124.in_EJ_R_heat_F_Yh %>%
+      filter(GCAM_region_ID %in% has_district_heat_GCAM_region_ID) %>%
+      left_join(L132.share_R_indenergy_heat_Yh, by = c("GCAM_region_ID", "year")) %>%
+      mutate(value = value.x * value.y) %>%
+      select(-fuel.y, -value.y, -value.x) %>%
+      rename(fuel = fuel.x) ->
+      L132.in_EJ_R_indheat_F_Yh
+
+    # Re-calculate industrial energy as original estimate minus fuel inputs to unconventional oil production, gas processing, and refining, and plus inputs to heat
+    L132.in_EJ_R_indenergy_F_Yh %>%
+      select(sector) %>%
+      unique %>%
+      unlist ->
+      temp_sector_value
+
+    L132.in_EJ_R_indenergy_F_Yh %>%
+      bind_rows(L132.in_EJ_R_inddeductions_F_Yh, L132.in_EJ_R_indheat_F_Yh) %>%
+      mutate(sector = temp_sector_value) ->
+      L132.in_EJ_R_Sindenergy_F_Yh
+
+    # Drop heat in regions where this fuel is backed out to its fuel inputs
+    A_regions %>%
+      filter(has_district_heat == 0) %>%
+      select(GCAM_region_ID) %>%
+      unique %>%
+      mutate(fuel = "heat") ->
+      region_heat
+
+    L132.in_EJ_R_Sindenergy_F_Yh %>%
+      anti_join(region_heat, by = c("GCAM_region_ID", "fuel")) %>%
+      group_by(GCAM_region_ID, sector, fuel, year) %>%
+      summarise(value = sum(value)) ->
+      ungroup ->
+      L132.in_EJ_R_indenergy_F_Yh
+
+    # Where unit conversions have produced slightly negative numbers that should be 0, re-set to 0
+    L132.in_EJ_R_indenergy_F_Yh %>%
+      filter(value < 0 & value > -1e-6) %>%
+      mutate(value = 0) ->
+      zero_value_rows
+
+    L132.in_EJ_R_indenergy_F_Yh %>%
+      filter(!(value < 0 & value > -1e-6)) %>%
+      bind_rows(zero_value_rows) %>%
+      ungroup ->
```

Comment:
Maybe do these two steps in one pipeline; also no need for `ungroup`?

Reply:
Without `ungroup` the driver throw out an error saying grouped data is not allowed. So I added the `ungroup` anyways and the driver wetn through... 

Also could you please give me a hint on how to do this in one pipeline? I thought about using `left_join` but that seems make things more complicated...

Reply:
I was thinking of something like this:
```R
    # Where unit conversions have produced slightly negative numbers that should be 0, re-set to 0
     L132.in_EJ_R_indenergy_F_Yh %>%
       filter(value < 0 & value > -1e-6) %>%
       mutate(value = 0) %>%
       bind_rows(L132.in_EJ_R_indenergy_F_Yh, filter(!(value < 0 & value > -1e-6)) ->
```
Up to you.

Re `ungroup`you've already ungrouped in line 166, so I didn't understand why it needed to be done again.

Reply:
Thanks Ben! @bpbond 

But for the `ungroup`, I'm still getting error running `driver`. I understand I `ungroup` the tibble before and my pipeline looks like this: 

```
    L132.in_EJ_R_indenergy_F_Yh %>%
      filter(value < 0 & value > -1e-6) %>%
      mutate(value = 0) %>%
      bind_rows(filter(L132.in_EJ_R_indenergy_F_Yh, !(value < 0 & value > -1e-6))) ->
      L132.in_EJ_R_indenergy_F_Yh
```
And the error from `driver` says:

```
[1] "module_energy_LA132.industry"
Error: L132.in_EJ_R_indenergy_F_Yh is being returned grouped. This is not allowed; please ungroup()
Called from: assert_that(is.null(groups(dots[[dname]])), msg = paste0(dname, 
    " is being returned grouped. This is not allowed; please ungroup()"))
```

I'm kind of confused...

Reply:
Odd. Well, leave it in then, no problem.